src/NServiceBus.Core/Pipeline/Outgoing/BatchOrImmediateDispatchConnector.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using DeliveryConstraints;
+    using Pipeline;
+    using TransportDispatch;
+    using Transports;
+
+    class BatchOrImmediateDispatchConnector : StageConnector<DispatchContext, ImmediateDispatchContext>
+    {
+        public async override Task Invoke(DispatchContext context, Func<ImmediateDispatchContext, Task> next)
+        {
+            var state = context.GetOrCreate<State>();
+            var dispatchConsistency = state.ImmediateDispatch ? DispatchConsistency.Isolated : DispatchConsistency.Default;
+
+            var options = new DispatchOptions(context.RoutingStrategy, dispatchConsistency, context.GetDeliveryConstraints());
+            var operation = new TransportOperation(context.Message, options);
+
+            PendingTransportOperations pendingOperations;
+
+            if (!state.ImmediateDispatch && context.TryGet(out pendingOperations))
```

Comment:
`pendingOperations` will not be there if I am sending directly from my bus instance, right? So we will plug it with an explicit send context and we'll be able to get rid of this additional condition.